File: 105.construct_binary_tree_from_preorder_and_inorder_traversal/buildTree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

# The basic idea is here:
# Say we have 2 arrays, PRE and IN.
# Preorder traversing implies that PRE[0] is the root node.
# Then we can find this PRE[0] in IN, say it's IN[5].
# Now we know that IN[5] is root, so we know that IN[0] - IN[4] is on the left side, IN[6] to the end is on the right side.
# Recursively doing this on subarrays, we can build a tree out of it :)


class Solution(object):
    def buildTree(self, preorder, inorder):
        """
        :type preorder: List[int]
        :type inorder: List[int]
        :rtype: TreeNode
        """
        return self.helper(0, 0, len(inorder)-1, preorder, inorder)

    def helper(self, preStart, inStart, inEnd, preorder, inorder):
        if preStart > len(preorder)-1 or inStart > inEnd:
            return None
        rootVal = preorder[preStart]
        root = TreeNode(rootVal)

        inIndex = inorder.index(rootVal)
        root.left = self.helper(preStart+1, inStart, inIndex-1, preorder,inorder)
        root.right = self.helper(preStart+inIndex-inStart+1, inIndex+1, inEnd, preorder, inorder)
        return root

File: 106.construct_binary_tree_from_inorder_and_postorder_traversal/buildTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {number[]} inorder
 * @param {number[]} postorder
 * @return {TreeNode}
 */

// function TreeNode(val) {
//     this.val = val;
//     this.left = this.right = null;
// }
var buildTree = function(inorder, postorder) {
    return helper(postorder.length-1, 0, inorder.length-1, inorder, postorder);
};

var helper = function(postIndex, inStart, inEnd, inorder, postorder){
    if (postIndex < 0 || inStart > inEnd) {
        return null;
    }
    var root = new TreeNode(postorder[postIndex]);
    var inIndex = inorder.indexOf(root.val);

    root.right = helper(postIndex-1, inIndex+1, inEnd, inorder, postorder);
    root.left = helper(postIndex-(inEnd-inIndex+1), inStart, inIndex-1, inorder, postorder);

    return root;
};

console.log(buildTree([4,2,5,1,6,3,7], [4,5,2,6,7,3,1]));